### PR TITLE
Change the ziputil dependency to fix a potential security concern

### PR DESCRIPTION
**Pull request by b4sjoo**
### Description
We've written our own ZipUtil class, where unzip method used Apache Commons’ ZipFile classes instead of ZipInputStream class. This can fix a potential security vulnerability.

### Issues Resolved
This resolved a sev3 ticket.
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Enumeration;
+
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * A util class contains file related operations.
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:19]: This java doc seems not so accurate, should be zip file related ?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+@Log4j2
+public class ZipUtils {
+    public static void unzip(File zipFile, Path dest) {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:23]: Add java doc for public method

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        try {
+            ZipFile file = new ZipFile(zipFile);
+            Enumeration<ZipArchiveEntry> en = file.getEntries();
+            ZipArchiveEntry ze;
+            while (en.hasMoreElements()) {
+                ze = en.nextElement();
+                String name = ze.getName();
+                if (name.contains("..")) {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:31]: What does this mean? 

Do you have any test case covering this branch? Same question for line 35

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                                "Parent path should never be null: " + f);
+                    }
+                    Files.createDirectories(parentFile);
+                    InputStream is = file.getInputStream(ze);
+                    Files.copy(is, f, StandardCopyOption.REPLACE_EXISTING);
+                    is.close();
+                }
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:50]: I just follow some samples here, changed it into `IllegalArgumentException` and specified error message.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java (hunk 3 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:31]: This is because previous djl unzip class wrote like this. I think it is for security concern but I cannot reproduce any file like that, so I cannot write a test case to cover it. Maybe delete it?

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        Files.createDirectories(output);
+        ZipUtils.unzip(file, output);
+    }
+
+    @Test
+    public void testUnzipFile() throws IOException, URISyntaxException {
+        File testZipFile = new File(Objects.requireNonNull(getClass().getResource("foo.zip")).toURI());
+        Path output = Paths.get("build/output");
+        Files.createDirectories(output);
+        ZipUtils.unzip(testZipFile, output);
```
[the remark this answers is not in the document]
Comment (b4sjoo) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java:34]: Added.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java (hunk 2 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:23]: Added.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java (hunk 1 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:19]: Corrected.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.engine.utils;
+
+import java.io.*;
```
[the remark this answers is not in the document]
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:8]: Corrected.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                                "Parent path should never be null: " + f);
+                    }
+                    Files.createDirectories(parentFile);
+                    InputStream is = file.getInputStream(ze);
+                    Files.copy(is, f, StandardCopyOption.REPLACE_EXISTING);
+                    is.close();
+                }
+            }
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Input file should never be null", e);
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:55]: Log the error?

Is "null file" is the only cause for IOException. How about make the error message more general ?

Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:55]: What about "Wrong input file"

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public static void unzip(File zipFile, Path dest) {
+        try {
+            ZipFile file = new ZipFile(zipFile);
+            Enumeration<ZipArchiveEntry> en = file.getEntries();
+            ZipArchiveEntry ze;
+            while (en.hasMoreElements()) {
+                ze = en.nextElement();
+                String name = ze.getName();
+                Path f = dest.resolve(name).toAbsolutePath();
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:39]: minor: generally we don't use single char as variable name except for some special case like `i` in for loop.  How about change to more readable name?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                if (!f.normalize().startsWith(dest))
+                    throw new RuntimeException("Bad zip entry");
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:41]: Can you write a test case to cover this branch?

Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:41]: Never mind, I guess the edge case should be like this one https://codeql.github.com/codeql-query-help/java/java-zipslip/#recommendation

If zip entry has name ` ..\sneaky-file` , then we may unzip to a parent folder.

File: ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        Path path = Paths.get("build/empty.zip");
+        File file = new File(path.toUri());
+        Path output = Paths.get("build/output");
+        Files.createDirectories(output);
+        ZipUtils.unzip(file, output);
+    }
+
+    @Test
+    public void testUnzipFile() throws IOException, URISyntaxException {
+        File testZipFile = new File(Objects.requireNonNull(getClass().getResource("foo.zip")).toURI());
```
Comment (dhrubo-os) [inline ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java:32]: We already have multiple zip files in the same resource folder. Do we need to add another zip file only for this test?

Comment (b4sjoo) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java:32]: I think the answer is yes because we cannot ensure these zip file will be there forever with same version or name.

Comment (dhrubo-os) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java:32]: For example, `all-MiniLM-L6-v2_onnx.zip` is being used in 3 difference tests. So I don't think this file will be disappear suddenly. Even if somebody replace this file, that will fail the test and they can update the changed file name accordingly.

In addition, in production this util will unzip a model zip file then why not use the similar file in test...

Comment (b4sjoo) [reply ml-algorithms/src/test/java/org/opensearch/ml/engine/utils/ZipUtilsTest.java:32]: Meanwhile I want to build a small size folder that has an empty sub-folder, a non-empty sub-folder and a file to cover most cases, I don't think we have that kind of zip file.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+ */
+@Log4j2
+public class ZipUtils {
+
+    /**
+     * Uncompressed a zip file.
+     * @param zipFile zip file to be uncompressed
+     * @param dest the destination path of this uncompress
+     */
+    public static void unzip(File zipFile, Path dest) {
```
[the remark this answers is not in the document]
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:31]: I think that's Ok because `mlEngine.getMlCachePath()` eventually goes to this `unzip` method as the `@dest` parameter

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java (hunk 8 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:39]: Corrected

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java (hunk 9 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/ZipUtils.java:41]: Yes but actually this edge case is really hard to create, linux does not allow me to do that...

**Review by ylwu-amzn — APPROVED** [review]

**Review by dhrubo-os — APPROVED** [review]
